c_src/bayesian.c

```diff
@@ -0,0 +1,208 @@
+        p->y = ei(y, sqrt(variance), fmin);
+
+        if (p->y > best) {
+            best = p->y;
+            for (size_t j=0; j<dims; ++j) { out[j] = p->x[j]; }
+        }
+
+        DataPoint* temp = points[oldest];
+        points[oldest] = p;
+        p = temp;
+
+        oldest = (oldest + 1) % EVO_POPULATION;
+
+    }
+
+    // Make sure to update the backing pointer array, since p may have been swapped
+    points[EVO_POPULATION] = p;
+}
+
+void shuffle_points(DataPoint** points, const size_t n) {
+    for (size_t i=0; i<n-1; ++i) {
+        const size_t j = i + random_u64() % (n-i);
+        DataPoint* temp = points[i];
+        points[i] = points[j];
+        points[j] = temp;
+    }
+}
+
+double bayesian_optimization(OptFunc f, const double* lower, const double* upper, double* x, const size_t dims,
+                             const size_t doe, const size_t iter) {
+
+    const size_t dp_size = sizeof(DataPoint) + dims * sizeof(double);
+
+    // Alloc backing data for all points (1 per iteration)
+    DataPoint* data = calloc(iter, dp_size);
+    
+    // Alloc pointers that can point into backing data points
+    DataPoint** points = calloc(iter, sizeof(DataPoint*));
+
+    // Assign each pointer to an allocated data point
+    points[0] = data;
+    for (size_t i=1; i<iter; ++i) { points[i] = (void*)points[i-1] + dp_size; }
+
+    // Alloc work space for local search algorithm
+    DataPoint* workspace = calloc(EVO_POPULATION + 1, dp_size);
+
+    // Pointers into workspace
+    DataPoint** workspace_points = calloc(EVO_POPULATION + 1, sizeof(DataPoint*));
+
+    workspace_points[0] = workspace;
+    for (size_t i=1; i<EVO_POPULATION+1; ++i) {
+        workspace_points[i] = (void*)workspace_points[i-1] + dp_size;
+    }
+
+    // This will correspond to the point in the parameter x
+    double best_val = INFINITY;
+
+    // Initialize DoE phase as a latin hypercube
+    for (size_t i=0; i<dims; ++i) {
+        for (size_t j=0; j<doe; ++j) {
+            points[j]->x[i] = lower[i] + (upper[i] - lower[i]) * (j + random_f64()) / doe;
+        }
+        shuffle_points(points, doe);
+    }
+
+    // Evaluate DoE points
+    for (size_t i=0; i<doe; ++i) {
+        const double y = f(points[i]->x);
+        points[i]->y = y;
+
+        if (y < best_val) {
+            best_val = y;
+            for (size_t j=0; j<dims; ++j) { x[j] = points[i]->x[j]; }
+        }
+    }
+
+    double mean_iter = 0.0;
+
+    for (size_t i=doe; i<iter; ++i) {
+
+        if (random_f64() < RANDOM_SAMPLE_PROBABILITY) {
+
+            get_random(dims, lower, upper, points[i]->x);
+
+        } else {
+            // Else get a guess from random forest model
+
+            // Fit forest
+            RandomForest forest;
+            randomforest_fit(&forest, NUMBER_OF_TREES, points, dims, i);
+
+            optimize_forest(&forest, lower, upper, points[i]->x, dims, best_val, workspace_points);
```

Comment:
What is this optimize forest part? Is it the acquisition function optimization?

Reply:
Yes!

Reply:
It'd be better to have a more standard BO naming here. Can we simply call it acquisition function or similar?

Reply:
Renamed optimize_forest() -> optimize_acquisition_function()